resources/templates/md/posts/2016-04-04-state-management.md

```diff
@@ -0,0 +1,364 @@
+2. Compare the current state of the `atom` and the state where the `swap` startet with transaction.
+3. If there's difference, make the transaction retry.
+4. If there's not difference, update the state.
+
+```clojure
+(defn number-of-marshmallows-left [marshmallows-eaten]
+  (swap! marshmallows-in-the-bag #(- % marshmallows-eaten)))
+```
+We created a `number-of-marshmalloes-left` function, which takes one argument `marshmallows-eaten`.
+In this function we passed `swap!` a function which takes one argument `current-state`.
+This function subtracts `marshmallows-eaten` from the current state.
+
+```clojure
+(number-of-marshmallows-left 10)
+
+;;=> 20
+```
+We can dereference the atom, and we will see that the `marshmallows-in-the-bag` value is updated
+
+```clojure
+@marshmallows-in-the-bag
+
+;;=> 20
+```
+
+## Agents
+
+`Agents` are similiar to `Atoms`, but in contrast with `atoms`, they are updated asynchronously.
+
+We need to keep in mind that the updates to agent will happen asynchronously and at unknown point in time, while updates in `atoms` will happened imediately.
+The `agents` are good for functions you want to get done, but it doesn't matter when they happen. 
+
+We will define `marshmallows-in-the-bag` as an agent.
+
+```clojure
+(def marshmallows-in-the-bag (agent 30))
+```
+
+To get the current value of the agent we will dereference it the same way as an `atom`. Using `clojure.core/deref` or `@` reader macro.
+
+```clojure
+@marshmallows-in-the-bag
+
+;;=> 30
+```
+
+To update the agent value (refers agent to the new state), we will use `send` or `send-off`.
+
+They are mostly similar, the difference is in how they're implemented.  
+
+If you send the actions to any agent using `send`, they run in a thread pool with a couple of more threads than the physsical number of processors. 
+If you make 1000 calls using `send`, the call that can't be processed imediately just wait until processor became available.
+If they block then the thread pool can run dry.
+
+If you send the actions to any agent using `send-off`, a new thread is created for each call. 
+if you make 1000 calls using `send-off`, the call that can't be processed immediately still wait for the next available processor, 
+but they may incur the extra overhead of starting a thread if the send-off threadpool happens to be running low. 
+It's ok if the threads block because each task (potentially) gets a dedicated thread.
+
+```clojure
+(defn number-of-marshmallows-left [marshmallows-eaten]
+  (send marshmallows-in-the-bag #(- % marshmallows-eaten)))
+```
+
+So if you call the `number-of-marshmallows-left` there's a possibility that the value is still not updated.
+
+```Clojure
+@marshmallows-in-the-bag
+
+;;=> 30
+```
+Few seconds later
+
+```Clojure
+@marshmallows-in-the-bag
+
+;;=> 20
+```
+
+
+## Refs
+
+The refs are ideal for managing states of more than one identity. 
+Especially when we want to update the state of more than one identity simultaneously.
```

Comment:
Well, I would delete this sentence as this is exactly the point we are trying to make.